spec/safe_statements_spec.rb

```diff
@@ -1692,24 +1692,116 @@ def up
+
+                expect(indexes.size).to eq(1)
+                expect(indexes.first).to have_attributes(
+                  table: table,
+                  name: "#{table}_updated_at_idx",
+                  columns: ["updated_at"],
+                  using: :btree,
+                )
+              end
+            end
           end
 
-          it "raises error when table is not partitioned" do
+          it "creates valid index when sub-partitioning detected and recurse is true" do
+            create_range_partitioned_table(:foos3, migration_klass, with_partman: true)
+            create_range_partitioned_table(:foos3_sub, migration_klass, with_partman: true)
+
             setup_migration = Class.new(migration_klass) do
               def up
-                safe_create_table :foos3 do |t|
-                  t.timestamps null: false
-                end
+                unsafe_execute(<<~SQL)
+                  ALTER TABLE foos3
+                  ATTACH PARTITION foos3_sub
+                  FOR VALUES FROM ('2020-01-01') TO ('2020-02-01')
+                SQL
+              end
+            end
+
+            setup_migration.suppress_messages { setup_migration.migrate(:up) }
+
+            test_migration = Class.new(migration_klass) do
+              def up
+                safe_add_concurrent_partitioned_index :foos3, :updated_at, name_suffix: "updated_at_idx", recurse: true
+              end
+            end
+
+            allow(ActiveRecord::Base.connection).to receive(:execute).and_call_original
+
+            test_migration.suppress_messages { test_migration.migrate(:up) }
+
+            aggregate_failures do
+              expect(ActiveRecord::Base.connection).to have_received(:execute).with(/CREATE INDEX "foos3_updated_at_idx" ON ONLY/).once
+              expect(ActiveRecord::Base.connection).to have_received(:execute).with(/CREATE INDEX "foos3_sub_updated_at_idx" ON ONLY/).once
+              expect(ActiveRecord::Base.connection).to have_received(:execute).with(/CREATE INDEX CONCURRENTLY/).exactly(20).times
+              expect(ActiveRecord::Base.connection).to have_received(:execute).with(/ALTER INDEX/).exactly(21).times
+            end
+
+            aggregate_failures do
+              # look up indexes on child tables, parent table, and sub-partitions
+              partitions_for_table(:foos3).append("foos3").concat(partitions_for_table(:foos3_sub)).each do |table|
+                indexes = ActiveRecord::Base.connection.indexes(table)
+
+                expect(indexes.size).to eq(1)
+                expect(indexes.first).to have_attributes(
+                  table: table,
+                  name: "#{table}_updated_at_idx",
+                  columns: ["updated_at"],
+                  using: :btree,
+                )
+              end
+            end
+          end
+
+          it "raises error when sub-partitioning detected and if_not_exists is false" do
```

Comment:
Why does sub-partitioning imply we must use `if_not_exists`?

Reply:
I think the test description is just wrong and it should be:

> raises error when sub-partitioning detected, recurse is false, and if_not_exists is false